In [1]:
!apt-get moo

                 (__) 
                 (oo) 
           /------\/ 
          / |    ||   
         *  /\---/\ 
            ~~   ~~   
..."Have you mooed today?"...


В этом ноутбуке мы рассмотрим базовые команды командной строки в Linux

План
  - Команды UNIX
  - Файловая система
  - Аргументы к командам
  - Переменные окружения


# Команды в UNIX

Линукс принадлежит большому семейству операционных систем под названием UNIX, коорые происходят от ОС AT&T Unix, разроботанной еще в 1970х. Mac OS, кстати тоже ее потомок.

В операционных системах UNIX-типа для взаимодействия с ОС используют специальные команды. До появления окошек и кнопочек люди создавали файлы и запускали команды исключительно с помощью этих команд. 

Сейчас этот способ тоже исользуют, тк есть ряд преимуществ:
 - Намного более удобное подключение к компьютеру по сети
 - Иногда удобнее запустить одну команду, чем бродить по менюшкам в поисках нужной опции.
 - Ломанды можно составлять в скрипты и автоматизировать задачи
 



Небольшой ликбез по понятиям:

- **Командная строка** - вид интерфейса, когда вы набираете текстовые команды
- **Терминал** - Программа, которая отображает строку. Например: *gnome-terminal*, *xterm*, *tilda*
- **Оболочка (shell)** - Программа которая выполняет команды. Например: *bash*, *zsh*, *ssh*
- **Консоль** - Железо, например клавиатура и монитор.

Запускать команды можно как из терминала, так и из юпитера или колаба, поставив перед командой !

In [2]:
!echo "Hello World"

Hello World


# Файловая система

В Unix системах все файлы образуют структуру похожую на дерево, в каждой директории(папке) хранятся файлы и другие директории. Самая первая директория которая содержит все остальные, называется корневой (*root*)


У каждой директории есть путь (по сути, просто строка) который показывает где она находится в этом дереве. "/" - это путь корневой директории. 

---



![alt text](http://linuxcommand.org/images/file_manager.jpg)

В командной строке все команды выполняются в какой-либо директории.

Представьте, что дерево файловой системы - это лабиринт, и каждая директория - это комната с файлами и дверьми в другие директории. В каждый момент времени вы находитесь в одной комнате.

Посмотрм, как нам ходить по нашему лабиринту.

In [3]:
# Выведем путь текущей директории
!pwd

/home/kl/Рабочий стол/DLS/advance/seminar


In [4]:
# Посмотрим что в нашей директории есть
!ls 
#ls is for "list"

'1.3 [seminar,basic]bash.ipynb'  '1.3 [seminar,basic]tools.ipynb'


Некоторые директории хранят файлы необходимые для работы программ. В них нужно что-либо менять только если значешь, что делаешь. 

Это так называемые скрытые директории (hidden или dot directories/files).

In [5]:
#

In [6]:
#Создадим пустой файлик
! touch test-file
# пробел в имени - плохая идея, создадутся два файла
! touch hello world

# посмотрим что получилось
! ls

'1.3 [seminar,basic]bash.ipynb'    hello       world
'1.3 [seminar,basic]tools.ipynb'   test-file


In [7]:
#Удаление файлов
! rm world
! ls

'1.3 [seminar,basic]bash.ipynb'    hello
'1.3 [seminar,basic]tools.ipynb'   test-file


In [9]:
# Новая директория
! mkdir please_come_here


Некоторые директории хранят файлы необходимые для работы программ. В них нужно что-либо менять только если значешь, что делаешь.

Это так называемые скрытые директории (hidden или dot directories/files).


In [10]:
# Покажем больше информации и отобразим скрытые файлы
!ls -al
# Первое слово - команда, а второе - аргумент

итого 72
drwxr-xr-x 4 kl kl  4096 окт  4 15:53  .
drwxr-xr-x 4 kl kl  4096 окт  4 14:13  ..
-rw-rw-r-- 1 kl kl 24611 окт  4 15:51 '1.3 [seminar,basic]bash.ipynb'
-rw-rw-r-- 1 kl kl 26696 окт  4 14:38 '1.3 [seminar,basic]tools.ipynb'
-rw-r--r-- 1 kl kl     0 окт  4 15:51  hello
drwxr-xr-x 2 kl kl  4096 окт  4 14:31  .ipynb_checkpoints
drwxr-xr-x 2 kl kl  4096 окт  4 15:53  please_come_here
-rw-r--r-- 1 kl kl     0 окт  4 15:51  test-file




```
.  - это путь к той директории в которой мы уже находимся
.. - это путь к родительской директории
.config - это скрытая папка
```

Каждая строчка - это одна из "дверей" в другую директорию либо файл.
*Относительный* путь это не полный "адрес" директории а то, куда нам нужно пойти из места в котором мы сейчас находимся. 


```
 drwxr-xr-x 1 root root 4096 Feb 15 17:21 please_come_here
 |                           дата создания    |
 |                                         _относительный_ путь
 |
  \ эта метка значит, что это путь к другой директории
```



In [13]:
# переход в другую директорию команда cd
! pwd
! cd please_come_here
! pwd

/home/kl/Рабочий стол/DLS/advance/seminar
/home/kl/Рабочий стол/DLS/advance/seminar


Заработало?

Не должно. Тут проблема в том что юпитер запускает каждый раз новую оболочку в своей директории (той, откуда его вызвали) , и после завершения команды она закрывается.

Чтоб команды запускались в новой директории нужно поменять директорию самого юпитера.


```
%cd
```



In [14]:
! pwd
% cd please_come_here
! pwd

/home/kl/Рабочий стол/DLS/advance/seminar
/home/kl/Рабочий стол/DLS/advance/seminar/please_come_here
/home/kl/Рабочий стол/DLS/advance/seminar/please_come_here


Когда мы переходим в другую директорию, ее имя добавляется к нашему пути через "/"

Например, были в `/home`, перешли в директорию `Pictures `оказались в `/home/Pictures`

Абсолютный путь файла в этой директории будет /home/Picutes/image.png 

Если мы хотим взять файл из директории, которая является дочерней к той директории, где мы сейчас находимся, мы можем указать относительный путь файла.

Например, из директории `/home` мы должны обращаться к нему по пути `Pictures/image.png`

In [15]:
# команды можно комбинировать в одну символом &
! ls & ls -a & pwd
%cd ..

# Чтоб команды выполнялись одна за одной нужно использовать &&
! pwd && cd please_come_here && pwd

/home/kl/Рабочий стол/DLS/advance/seminar/please_come_here
/home/kl/Рабочий стол/DLS/advance/seminar
/home/kl/Рабочий стол/DLS/advance/seminar
/home/kl/Рабочий стол/DLS/advance/seminar/please_come_here


## Запись и чтение в файлы

In [16]:
# Вывести текст в консоль

!echo "Sample text Пример текста"
# Echo - потому что команда выводит на выход то же, что и получает

Sample text Пример текста


In [17]:
# Вместо консоли отправить в файл

! echo "London is the capital of Great Britain" >> English.txt


In [18]:
# Прочитать файл и вывести в консоль

! cat English.txt

London is the capital of Great Britain


## Скачивание из интернета

In [19]:
# сохраняем старницу по адресу

! curl example.com >> example-com.html
! ls
! cat example-com.html


  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1256  100  1256    0     0   5277      0 --:--:-- --:--:-- --:--:--  5277
'1.3 [seminar,basic]bash.ipynb'    example-com.html   test-file
'1.3 [seminar,basic]tools.ipynb'   hello
 English.txt			   please_come_here
<!doctype html>
<html>
<head>
    <title>Example Domain</title>

    <meta charset="utf-8" />
    <meta http-equiv="Content-type" content="text/html; charset=utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type="text/css">
    body {
        background-color: #f0f0f2;
        margin: 0;
        padding: 0;
        font-family: -apple-system, system-ui, BlinkMacSystemFont, "Segoe UI", "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
        
    }
    div {
        width: 600px;
        margin: 5em auto;
        padding: 2em;
        background-color: #fdfdf

In [20]:
! curl instagram.com

Иногда сервера пытаются перенаправить запрос на другой адрес. Curl не следует по таким перенаправлениям. Есть прогаммка wget которая может переходить по таким ссылкам и вообще проходить по всем ссылкам на странице и скачивать их.

In [21]:
! wget instagram.com -O insta.html

--2020-10-04 15:58:12--  http://instagram.com/
Распознаётся instagram.com (instagram.com)… 52.1.238.189, 35.172.167.44, 3.226.111.93, ...
Подключение к instagram.com (instagram.com)|52.1.238.189|:80... соединение установлено.
HTTP-запрос отправлен. Ожидание ответа… 301 Moved Permanently
Адрес: https://instagram.com/ [переход]
--2020-10-04 15:58:13--  https://instagram.com/
Подключение к instagram.com (instagram.com)|52.1.238.189|:443... соединение установлено.
HTTP-запрос отправлен. Ожидание ответа… 301 Moved Permanently
Адрес: https://www.instagram.com/ [переход]
--2020-10-04 15:58:13--  https://www.instagram.com/
Распознаётся www.instagram.com (www.instagram.com)… 31.13.72.174, 2a03:2880:f20a:e5:face:b00c:0:4420
Подключение к www.instagram.com (www.instagram.com)|31.13.72.174|:443... соединение установлено.
HTTP-запрос отправлен. Ожидание ответа… 200 OK
Кукисы, полученные из www.instagram.com, попытались изменить домен на i.instagram.com
Кукисы, полученные из www.instagram.com, попыт

Обратите внимание на

`HTTP request sent, awaiting response... 301 Moved Permanently`

# Аргументы и параметры к командам

Обычно аргументы – это то, над чем выполняется команда. Например, команда перехода в другую директорию должна получить в качестве аргумента путь директории, куда следует перейти. Команде вывода на экран содержимого файла надо передать имя файла.

Параметры модифицируют, как бы видоизменяют работу программы. Например, с помощью команды просмотра содержимого каталога можно вывести простой список вложенных объектов. Используя же определенный параметр, можно получить подробные сведения о каждом файле и папке.

Аргументы идут через пробел как отдельные слова

Параметры идут после дефиса: ` -f` (однобуквенные) либо после двух дефисов: `--file`  

In [25]:
# календарик на весь год

!ncal 1997


                                  1997
    Января            Февраля           Марта             Апреля            
Пн     6 13 20 27        3 10 17 24        3 10 17 24 31     7 14 21 28   
Вт     7 14 21 28        4 11 18 25        4 11 18 25     1  8 15 22 29   
Ср  1  8 15 22 29        5 12 19 26        5 12 19 26     2  9 16 23 30   
Чт  2  9 16 23 30        6 13 20 27        6 13 20 27     3 10 17 24      
Пт  3 10 17 24 31        7 14 21 28        7 14 21 28     4 11 18 25      
Сб  4 11 18 25        1  8 15 22        1  8 15 22 29     5 12 19 26      
Вс  5 12 19 26        2  9 16 23        2  9 16 23 30     6 13 20 27      

    Мая               Июня              Июля              Августа           
Пн     5 12 19 26        2  9 16 23 30     7 14 21 28        4 11 18 25   
Вт     6 13 20 27        3 10 17 24     1  8 15 22 29        5 12 19 26   
Ср     7 14 21 28        4 11 18 25     2  9 16 23 30        6 13 20 27   
Чт  1  8 15 22 29        5 12 19 26     3 10 17 24 31   

In [26]:
# передаем 2 аргумента

!ncal 03 2019


    Марта 2019        
Пн     4 11 18 25   
Вт     5 12 19 26   
Ср     6 13 20 27   
Чт     7 14 21 28   
Пт  1  8 15 22 29   
Сб  2  9 16 23 30   
Вс  3 10 17 24 31   


In [27]:
# передаем параметр -b

!ncal -b 03 2019


     Марта 2019       
Пн Вт Ср Чт Пт Сб Вс  
             1  2  3  
 4  5  6  7  8  9 10  
11 12 13 14 15 16 17  
18 19 20 21 22 23 24  
25 26 27 28 29 30 31  
                      


In [28]:
# Однобуквенные параметры можно склеивать

!ncal -b -w 03 2019
!ncal -bw 03 2019

       Марта 2019          
 w| Пн Вт Ср Чт Пт Сб Вс   
 8|              1  2  3   
 9|  4  5  6  7  8  9 10   
10| 11 12 13 14 15 16 17   
11| 18 19 20 21 22 23 24   
12| 25 26 27 28 29 30 31   
                           
       Марта 2019          
 w| Пн Вт Ср Чт Пт Сб Вс   
 8|              1  2  3   
 9|  4  5  6  7  8  9 10   
10| 11 12 13 14 15 16 17   
11| 18 19 20 21 22 23 24   
12| 25 26 27 28 29 30 31   
                           


# Переменные окружения

Оболочка в себе содержит так называемые переменные окружения - глобальные константы которые доступны всем программам которые она выполняет. В них удобно хранить например путь к конфигурационным файлам или порт какого нибудь сервиса.

Переменные окружения зачастую определяют как ведет себя программа. Например, питон ищет пакеты в переменной `PYTHONPATH`

В переменной `PATH` находятся пути в которых Bash ищет исполняемые файлы для других программ


In [29]:
# обьявляем переменную и считываем ее ($ перед именем)
!FOO=kindersurprise && echo $FOO

kindersurprise


In [30]:
!echo $PATH
!echo $PYTHONPATH

/home/kl/anaconda3/bin:/home/kl/anaconda3/bin:/home/kl/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin



Мы можем создать новую папку и заставить питон думать что в ней лежит модуль

In [34]:
! rm -r fakemodule
!mkdir fakemodule
!mkdir fakemodule/power
! echo "print('A great power comes with a great responsibility')" >> fakemodule/power/__init__.py




In [36]:
!export a="Эта переменная будет доступна другим программам вызванным в той же оболочке" && echo $a
# Но тут уже не получится, для каждой команды Jupyter запускает свою оболочку 
! echo $a

Эта переменная будет доступна другим программам вызванным в той же оболочке



In [0]:
! PYTHONPATH=$PYTHONPATH:$(pwd)/fakemodule && python -c 'import power'

конструкция `$(pwd)` подставляет вывод команды `pwd` в выражение. 


```
PYTHONPATH=$PYTHONPATH:$(pwd)/fakemodule

```
После подстановки переменных мы получим 

```
PYTHONPATH=<то что было в PYTHONPATH>:<текущая директория>/fakemodule

```
Это добавит нашу fakemodule в переменную окружения и позволит питону найти наш модуль



In [38]:
!rm -r fakemodule